feat(ssi): add namespace support for labels/expressions

betterengineering opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
This commit adds the ability to use a namespace selector to match against namespaces defined in the targets list. For example, this config will match a pod in a namespace with `tracing: yes` and `env: prod`:
```yaml
    - name: "Enabled Prod Namespaces"
        namespaceSelector:
          matchLabels:
            tracing: "yes"
          matchExpressions:
            - key: "env"
              operator: "In"
              values:
                - "prod"
        ddTraceVersions:
          python: "default" 
```

This PR maintains the ability to use `matchNames` as a convenience method. For example, this config would match the billing-service namespace:
```yaml
      - name: "Billing Service"
        namespaceSelector:
          matchNames:
          - "billing-service"
        ddTraceVersions:
          python: "default"       
```

### Motivation
This change is adding the full support for namespaces defined in [Kubernetes SSI | Workload Selection 🎯](https://docs.google.com/document/d/1Lol1ExLF1nGBe7njO6DBVkjuYAYxC1-sL2WP0rdqFRk/edit?usp=sharing). It was not added to the change in [feat(ssi): add target based filtering](https://github.com/DataDog/datadog-agent/pull/33561) as it required a bit more effort to get correct.

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->
The added unit tests offer strong coverage of these changes. To ensure they worked outside of unit tests, I also ran the following tests.

#### Metadata collection
To ensure that the metadata collection for namespaces is enabled, I used the following Helm config:
```
  datadog_cluster_yaml:
    apm_config:
      instrumentation:
        enabled: true
        targets:
          - name: "Microservices"
            selector:
              matchLabels:
                language: "python"
            ddTraceVersions:
              python: "default"
```

I ran the following:
```
 kubectl -n system exec -it svc/datadog-agent-cluster-agent -c cluster-agent -- agent workload-list
```

Here is a snippet of the output with my namespaces defined:
```
...
=== Entity kubernetes_metadata sources(merged):[kubeapiserver] id: /namespaces//system ===
----------- Entity ID -----------
Kind: kubernetes_metadata ID: /namespaces//system

----------- Entity Meta -----------
Name: system
Namespace:
===

=== Entity kubernetes_metadata sources(merged):[kubeapiserver] id: /namespaces//test-1 ===
----------- Entity ID -----------
Kind: kubernetes_metadata ID: /namespaces//test-1

----------- Entity Meta -----------
Name: test-1
Namespace:
===
...
```

#### Injector metadata
To ensure I was able to fetch the metadata for a namespace during injection, I added temporary code to the inject function:
```
func (w *Webhook) inject(pod *corev1.Pod, ns string, cl dynamic.Interface) (bool, error) {
	id := util.GenerateKubeMetadataEntityID("", "namespaces", "", pod.Namespace)
	namespace, err := w.wmeta.GetKubernetesMetadata(id)
	if err != nil {
		return false, fmt.Errorf("[MARK HACKS] could not get kubernetes namespace to match against for %s: %w", pod.Namespace, err)
	} else {
		log.Info("[MARK HACKS] namespace found from metadata: %v", namespace)
	}

	return w.mutator.MutatePod(pod, ns, cl)
}
```
After deploying a pod that received injection, I saw the following log line:
```
│ cluster-agent 2025-02-09 14:29:35 UTC | CLUSTER | INFO | (pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:132 in inject) | [MARK HACKS] namespace found from metadata: %v &{{kubernetes_metadata /namespaces//test-1} {test-1  map[kubectl.kubernetes.io/last-applied-configuration:{"apiVersion":"v1","kind":"Namespace","metadata":{"annotations":{},"name":"test-1"}}
```

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->
One thing to note, this config/filter is not yet active or exposed. It's simply modifying a struct with unit tests that's not wired in anywhere. We're working hard to get this feature ready for the next agent release milestone and will include release notes, additional testing, etc before that happens.

betterengineering committed:
feat(ssi): add namespace support for labels/expressions

This commit adds the ability to use a namespace selector to match
against namespaces defined in the targets list.

betterengineering commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_filter.go:
```diff
@@ -103,13 +127,31 @@ func (f *TargetFilter) filter(pod *corev1.Pod) []libInfo {
+		// Get the namespace metadata. At the time of writing, this method will only return an error if the namespace
+		// does not exist, in which case we return false for this selector.
+		id := util.GenerateKubeMetadataEntityID("", "namespaces", "", namespace)
+		meta, _ := t.wmeta.GetKubernetesMetadata(id)
```
@davidor is this the best way to do this? How does this store work? Is it safe to assume the namespace will be there by the time we get a mutating webhook?

davidor replied:
The call is correct, but there's something to take into account.

The KubernetesMetadata objects are collected by the `kubeapiserver` workloadmeta collector. The code is here: `comp/core/workloadmeta/collectors/internal/kubeapiserver`. However, by default, the collector only gets metadata for nodes. For other resources, like pods and deployments it enables the collection depending on some configuration options. We will need a condition there to enable namespace metadata collection according to some APM SSI configuration option.

For your local tests, you can run the `agent workload-list` command on the cluster-agent pod to see what is being collected.

There is no guarantee that the namespace will be available when this code runs. There will always be a slight delay between the creation or deletion of a namespace in Kubernetes and its storage in workloadmeta. So it is possible that the injection might not happen immediately after a namespace is created in Kubernetes. However, if everything works as expected, the lag should be minimal.

davidor commented on pkg/clusteragent/admission/mutate/autoinstrumentation/target_filter_test.go:
```diff
@@ -148,8 +215,23 @@ func TestTargetFilter(t *testing.T) {
+			// Create a mock meta.
+			wmeta := fxutil.Test[workloadmetamock.Mock](t, fx.Options(
+				fx.Supply(coreconfig.Params{}),
+				fx.Supply(log.Params{}),
```
Not sure if this `fx.Supply(log.Params{})` is needed given that just below have `fx.Provide(func() log.Component { return logmock.New(t) })`. Can you check?

I'm also not sure about `fx.Supply(context.Background())` I have not seen in other tests.

davidor commented on pkg/clusteragent/admission/mutate/autoinstrumentation/testdata/filter.yaml:
```diff
@@ -21,6 +21,17 @@ apm_config:
+        namespaceSelector:
+          matchLabels:
+            tracing: "yes"
+          matchExpressions:
```
Now that I see this, I'm not sure if `matchExpressions` is a good name.
Is it clear that it refers to labels? Maybe this is something that we can discuss later on when we review all the new config options.

betterengineering replied:
MatchLabels/MatchExpressions are [a standard Kubernetes concept](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#service-and-replicationcontroller). Straying from that would likely add more confusion.

betterengineering committed:
Remove additional fx providers

betterengineering committed:
feat(ssi): enable namespace collection if ssi is enabled

This commit adds namespace metadata collection if auto instrumentation
is enabled with a target list defined.

betterengineering committed:
Merge branch 'main' into mark.spicer/INPLAT-466-add-ns-matchlabels-support

davidor commented on comp/core/workloadmeta/collectors/internal/kubeapiserver/kubeapiserver.go:
```diff
@@ -146,6 +153,34 @@ func resourcesWithExplicitMetadataCollectionEnabled(cfg config.Reader) []string
+	// Targets is a custom struct type, so we unmarshal it into an interface
+	// slice to avoid the import while still being able to check if it's empty.
+	targets := []interface{}{}
+	err := cfg.UnmarshalKey("apm_config.instrumentation.targets", &targets)
```
Shouldn't we be more specific here and check for the namespace selector that needs the namespace labels?

betterengineering replied:
In order to support that, `kubeapiserver` would need to depend on `autoinstrumentation` which felt wrong as an import. It also felt a bit more complicated to describe to customers and test. But if you think it's worth it as a trade off, I'm happy to add the import so we can parse the target object.

davidor replied:
Importing extra dependencies can be problematic, depending on which agent requires them and whether they increase the binary size. We have CI checks in place for this.

Fixing this might not be worth it, but I think that at least we should add a comment here indicating that for simplicity, we collect namespace metadata if targets are defined, even if those targets only refer to pods rather than namespaces. This is something that can be optimized later if needed.

betterengineering committed:
Fix log import

betterengineering committed:
Add comment about only checking if targets exist
